Split up LPMS HTTP listener from CLI

j0sh opened the pull request:
Lots of user facing API changes within here, so take note. Many of these choices are rather arbitrary, so we can change up options and defaults however folks prefer.

* CLI gets its own HTTP mux, but the default listener remains on `127.0.0.1:8935`. This can be configured via `-cliAddr ip:port` .
* The default port for LPMS becomes 7935. PR for LPMS [here](https://github.com/livepeer/lpms/pull/85), while the lpms-goclient merge is at https://github.com/livepeer/go-livepeer/pull/507/commits/34a7f4733ae2aa05e0134a9d86b308bb8679d20b . However, **we don't use the LPMS default in goclient**.
* In goclient, we set the broadcaster's LPMS HTTP port for to 4433 (rather than the default of 7935) to match the transcoder's default RPC port. The transcoder's LPMS mux is shared with RPC.
* The RPC/HTTP listener can be configured with `-httpAddr ip:port` . This is a merge of the separate `-httpIP` and `-httpPort` options.
* For the broadcaster's default LPMS/HTTP address is 127.0.0.1:4433 unless configured via `-httpAddr`. Note that the broadcaster will be serving HTTP, not HTTPS.
* The transcoder's default RPC/HTTP address is `:<serviceURI Port>` unless configured via -httpAddr. Listens to all interfaces.
* The `-publicAddr` is renamed to `-serviceAddr` in an attempt at naming consistency with the Service URI. (It's not actually named `'-serviceURI` since it's not a complete URI, just an `ip:port` pair. I'll probably regret this one day.) The address check was also expanded to include the port.
* Note that the serviceAddr is a sanity check: it ensures that the transcoder's expectation of its service URI is consistent with what is actually on the blockchain. If those differ, the transcoder is not getting any jobs. For now, though, we just warn if there's a mismatch. Eventually we should halt the node on a mismatch.

Will update the documentation with an elaboration on the various addressing options, but if there's anything about the above explanation(s) that are unclear, please let me know.

j0sh committed:
transcoder: Return a value after transcoding.

j0sh committed:
mediaserver: Share HTTP mux between LPMS and RPC.

j0sh committed:
cli: Make server use non-default mux.

Because we can elect to use the default mux in LPMS sometimes.

j0sh committed:
livepeer: Make LPMS HTTP address configurable.

j0sh committed:
rpc: Generate code for transcoder responses.

j0sh committed:
rpc: Send transcoded results back to client.

j0sh commented:
Note that the broadcaster now uses the existing -publicAddr to configure the LPMS http port.

This is subject to change: in the transcoder, -publicAddr is meant to be the address that's used to match the ServiceURI. Might add another -httpAddr option instead, and use that for configuring the listen interfaces for the broadcaster/transcoder (while keeping -publicAddr as-is to check against the Service URI). That would be conceptually cleaner, but I'd also like to avoid overloading the user with options. Might not be so bad if the defaults are reasonable and it's thoroughly explained in the documentation.

Thoughts?

j0sh committed:
transcoder: Simplify the internal API.

j0sh committed:
claimmanager: Return transcoded data hash.

j0sh committed:
monitor: Accommodate new transcode response format.

j0sh committed:
broadcaster: Download segments from the transcoder.

j0sh committed:
livepeer: Split Service URI config from RPC listener.

j0sh committed:
proto: Move protobufs definitions into net.

Allows us to use the data types outside the server package.

j0sh committed:
proto: Update references to new proto location.

j0sh committed:
vendor: Update LPMS to d8a9224.

ericxtang commented on cmd/livepeer/livepeer.go:
```diff
@@ -77,7 +77,7 @@ func main() {
 	noIPFSLogFiles := flag.Bool("noIPFSLogFiles", false, "Set to true if log files should not be generated")
 	offchain := flag.Bool("offchain", false, "Set to true to start the node in offchain mode")
-	publicAddr := flag.String("publicAddr", "", "Public address that broadcasters can use to contact this node; may be an IP or hostname. If used, should match the on-chain ServiceURI set via livepeer_cli")
+	serviceAddr := flag.String("serviceAddr", "", "Transcoder only. Public address:port that broadcasters can use to contact this node; may be an IP or hostname. If used, should match the on-chain ServiceURI set via livepeer_cli")
```
I think we should match the name used in the protocol here and change to `ServiceURI` - and we should make it an actual URI. The right thing to do is probably to encourage transcoders to set up `https`, even though it adds some operational complexity for the transcoders. 

About the port number, is there a reason we are switching to a new port `4433`?  Seems like the least impact for an end user would be to:
* Keep the default http/rpc port as 8935
* Keep the default rtmp port as 1935
* Move the default cli port to something else (for example, 7935), and have it set to be the same default in the cli.

This way, if the users are using default ports, they won't even notice the port changes.

j0sh replied:
Yeah, there's a few somewhat conflicting goals here. 

* We already force HTTPS in the Service URI; it's added in implicitly by the CLI. The user doesn't have to specify a the `https` scheme when setting the ServiceURI, just ip:port . See https://github.com/livepeer/go-livepeer/blob/df3618534c37a4ef03ea1608e97ba7a9b0b468fc/cmd/livepeer_cli/wizard_transcoder.go#L47-L62
* `-serviceAddr` was used to try and maintain consistency with the `xxxAddr` scheme of the other options. We have rtmpAddr, httpAddr and cliAddr . Those only require an ip:addr , so we only require an ip:addr when setting serviceAddr. (This also matches how the CLI asks for the service URI before setting it on-chain.) Internally, we construct that into a full URI and parse that during the check at startup.
* The default ServiceURI port is already 4433. (At the time, we couldn't reuse 8935.) We can change this to 8935. Note that the 'https alternative' port, 8443, may be more friendly to some firewalls, if we're going to make a change. https://github.com/livepeer/go-livepeer/blob/df3618534c37a4ef03ea1608e97ba7a9b0b468fc/cmd/livepeer_cli/wizard_transcoder.go#L14
* Users will notice the CLI port changes if they are invoking the CLI endpoints (`/debug`, `/status` , `/manifestID`).

Given those choices, I think I'd rather keep the CLI port as-is, and maybe change the default RPC/HTTP port to 8443. That should buy us a little more compatibility with broadcasters.

Whatever we do, we just have to stress this difference in the documentation. Mostly that means communicating to our existing users. I can post a thorough explanation on the forums and the docs. Hopefully newcomers won't have trouble picking this up, since they don't have prior conditioning to break.

ericxtang replied:
I get your points.  All sounds good.

Re: CLI port vs. RPC/HTTP port, I'm just thinking about the changes that needs to be done for current transcoder.  If the RPC/HTTP port changes, they would have to change their AWS security group settings, which is kind of a pain.  If the CLI port changes, it'l most likely an change for things done on localhost, which is less of an issue.  Thoughts?

j0sh replied:
Sure, set the default CLI to 7935 and RPC/HTTP to 8935 . https://github.com/livepeer/go-livepeer/pull/512/commits/d8223b265031242b1cb43f3260a1cc83d23769a8

ericxtang commented on cmd/livepeer/livepeer.go:
```diff
@@ -54,11 +54,11 @@ func main() {
-	httpIP := flag.String("httpIP", "127.0.0.1", "IP to bind for HTTP RPC commands")
+	rtmpIP := flag.String("rtmpIP", "127.0.0.1", "IP to bind for RTMP commands")
+	cliAddr := flag.String("cliAddr", "127.0.0.1:8935", "Address to bind for  CLI commands")
+	httpAddr := flag.String("httpAddr", "", "Address to bind for HTTP commands")
```
If `httpAddr` will be the same as `serviceAddr` (or `serviceURI`), can we just get rid of it?

j0sh replied:
Those won't always be the same.

* The serviceAddr indicates the ip:port that is stored on the blockchain. This must match what is on the blockchain. (This is a 'soft must' for now.) If unset, the default is behavior to auto-discover the node's current public IP and check that against the blockchain.  There's a bunch of reasons someone would want to set the serviceAddr: if the serviceURI is a hostname, if running a private livepeer network where the serviceURI is non-public, etc. This option is really a sanity check to ensure that the transcoder's ServiceURI expectations match what is actually on the blockchain. If there's an unhandled mismatch, then it's likely the transcoder would not be able to actually receive any jobs.
* If the httpAddr is unset, listen to the service port that is stored on-chain.
* However, if the orchestrator's listening interface needs to differ from the serviceURI, then httpAddr is set. For example, if the node traffic is port-forwarded or load-balanced. So httpAddr is set in order to configure the listening interface.
* Broadcaster still needs a `-httpAddr` option since `-serviceAddr` doesn't have any meaning for them.

Most transcoders will not need to set either of these options, at least early on when things are still simple.

Writing this up I think I found an issue with the code, so will investigate that in the AM.

ericxtang commented on cmd/livepeer/livepeer.go:
```diff
@@ -54,11 +54,11 @@ func main() {
 	datadir := flag.String("datadir", fmt.Sprintf("%v/.lpData", usr.HomeDir), "data directory")
-	rtmpIP := flag.String("rtmpIP", "127.0.0.1", "IP to bind for HTTP RPC commands")
-	httpIP := flag.String("httpIP", "127.0.0.1", "IP to bind for HTTP RPC commands")
+	rtmpIP := flag.String("rtmpIP", "127.0.0.1", "IP to bind for RTMP commands")
```
For consistency sake, should we combine `rtmpIP` and `rtmpPort` into `rtmpAddr`?  (same as `cliAddr`)

j0sh replied:
Yep, done.

j0sh replied:
https://github.com/livepeer/go-livepeer/pull/507/commits/0d71cbb119561d8ec89227098b60af9bfca629db

ericxtang reviewed (commented):
Left some comments

j0sh committed:
core: Fix tests.

j0sh committed:
server: Fix tests.

j0sh committed:
eth: Fix tests.

j0sh committed:
livepeer: Merge rtmpIp and rtmpPort into rtmpAddr.

j0sh committed:
rpc: Privatize getTranscoder.

j0sh committed:
transcoder: Cosmetic renames.

j0sh committed:
videocache: Flatten VideoNetwork.

This also leaves VideoNetwork with no significant purpose, so fold
the remaining into VideoCache and remove VideoNetwork.

j0sh committed:
core: Additional test cleanups.

j0sh committed:
transcoder: Move GetClaimManager into transcoder.go.

Also document the logical layout of the code within LivepeerNode.

j0sh committed:
orchestrator: Rename file.

j0sh committed:
videosource: Rename from VideoCache.

j0sh committed:
rpc: Fix crash on non-200 response.

j0sh committed:
core: Remove the notion of a NodeID.

Also add some tests around StreamID / ManifestID.

j0sh committed:
livepeer: Switch default ports around.

HTTP/RPC becomes 8935
CLI becomes 7935

j0sh committed:
livepeer: Use blockchain ServiceURI rather than locally inferred URI.

darkdarkdragon committed:
Using blocking connection to the rpc server - this will give us clear error in case transcoder is down

j0sh committed:
core: Length checks for ManifestID/StreamID.

Prevents crashes on malformed input.

j0sh committed:
jobwatcher: Make log parsing more robust.

Prevents crashes with malformed StreamIDs.

j0sh committed:
jobwatcher: Retry the assigned transcoder lookup.

More robust against sporadic Eth RPC failures.

j0sh committed:
rpc: Misc fixes.

j0sh committed:
README: Update.

ericxtang committed:
Merge pull request #513 from livepeer/ja/nodeid

Remove Node ID

yondonfu committed:
Merge pull request #516 from livepeer/it/errors_logging

Better errors reporting

yondonfu committed:
Merge pull request #512 from livepeer/ja/v2followup

Follow up changes to v2 feedback

yondonfu committed:
Merge pull request #511 from livepeer/ja/fixtests

Fix tests

yondonfu committed:
Merge pull request #508 from livepeer/ja/tcresponse

Structured responses from transcoder

yondonfu reviewed (approved):
🚢
